File: tests/host/security_policy_target_tests.py
"""Compile actual policy console with SDK seams; no target source substitution."""
from pathlib import Path
import shutil,subprocess,tempfile
import receipt_console_binding_tests as receipt
import usb_console_binding_tests as base
ROOT=Path(__file__).resolve().parents[2]
def run():
    compiler=shutil.which("g++")
    if compiler is None:raise RuntimeError("native g++ required")
    with tempfile.TemporaryDirectory(prefix="policy-console-") as d:
        work=Path(d)
        stubs=dict(receipt.STUBS)
        stubs["hal/usb_serial_jtag_ll.h"]+="bool usb_serial_jtag_ll_rxfifo_data_available();\nunsigned usb_serial_jtag_ll_read_rxfifo(std::uint8_t*,unsigned);\n"
        for name,text in stubs.items():
            p=work/name;p.parent.mkdir(parents=True,exist_ok=True);p.write_text(text)
        harness=base.HARNESS+"\nvoid esp_rom_delay_us(unsigned n){check(n>0&&n<=5);clock_value+=n;}\nstd::string rx;bool usb_serial_jtag_ll_rxfifo_data_available(){return !rx.empty();}\nunsigned usb_serial_jtag_ll_read_rxfifo(std::uint8_t* p,unsigned n){check(n==1&&!rx.empty());*p=rx[0];rx.erase(0,1);return 1;}\n"
        source=ROOT/"firmware/targets/heltec_v4_security_policy_eval/main/policy_console.cpp"
        cpp=work/"actual.cpp";exe=work/"actual.exe"
        cpp.write_text(harness+'\n#include "'+source.as_posix()+'"\n'+(ROOT/"tests/host/security_policy_target_tests.cpp").read_text())
        subprocess.run([compiler,"-std=c++17","-Wall","-Wextra","-Werror","-I",str(work),"-I",str(ROOT/"firmware/components/diagnostics/include"),"-I",str(ROOT/"firmware/components/security_evaluation/include"),"-I",str(ROOT/"firmware/targets/heltec_v4_security_policy_eval/main"),str(cpp),"-o",str(exe)],check=True,timeout=90)
        for scenario in ["disabled","read","success","stalled","late","bypass","loop","loop_cap","loop_invalid","stdio"]:
            result=subprocess.run([str(exe),scenario],check=True,capture_output=True,text=True,timeout=10)
            if result.stdout.strip()!=f"PASS {scenario}":raise RuntimeError("unexpected output")
        print("PASS 10 actual policy console/control groups")
if __name__=="__main__":run()

File: tests/host/security_policy_target_tests.cpp
#include "policy_control_loop.hpp"
int main(int argc,char**argv){
 check(argc==2);std::string scenario=argv[1];
 check(__esp_system_init_fn_ot_receipt_quarantine()==ESP_OK);check(ot_console_install());
 if(scenario=="loop" || scenario=="loop_cap" || scenario=="loop_invalid"){
  using namespace opentrail::security_evaluation;
  Control c(0);unsigned reads=0,pauses=0;std::uint64_t time=1;
  const std::string command="RUN SEC_EVAL1 ot187-policy-v0 0123456789abcdef0123456789abcdef\n";
  std::string input=scenario=="loop"?command+command:scenario=="loop_invalid"?"bad\n": "";
  auto read=[&](char& b){++reads;if(input.empty())return false;b=input[0];input.erase(0,1);return true;};
  auto clock=[&]{return time;};auto pause=[&]{++pauses;};auto healthy=[]{return true;};
  bool ok=receive_control(c,read,clock,pause,healthy);
  if(scenario=="loop"){
   check(ok&&input==command&&reads==command.size());
   check(!receive_control(c,read,clock,pause,healthy)&&input==command&&reads==command.size());
  }else if(scenario=="loop_cap")check(!ok&&pauses==7000&&reads==7000);
  else check(!ok&&c.state()==State::refused&&reads==4);
  std::cout<<"PASS "<<scenario<<"\n";return 0;
 }
 const std::string receipt="SEC_EVAL1 ot187-policy-v0 0123456789abcdef0123456789abcdef pass\n";
 if(scenario=="disabled"){check(!ot_policy_send(receipt.data(),receipt.size()));check(accepted_bytes==0);}
 else if(scenario=="read"){rx="ABC";char byte=0;check(ot_policy_read(byte)&&byte=='A');check(ot_policy_read(byte)&&byte=='B');check(ot_policy_read(byte)&&byte=='C');check(!ot_policy_read(byte));}
 else{
 check(ot_console_begin_session());
 check(!ot_console_begin_session());
 if(scenario=="stalled")stall_after_write=true;
 if(scenario=="late")overrun_write=true;
 if(scenario=="bypass")installed[0]('x');
 if(scenario=="stdio"){_reent r;check(__wrap__write_r(&r,1,"x",1)==-1&&r.error==EIO);}
 const bool ok=ot_policy_send(receipt.data(),receipt.size());
 if(scenario=="success"){
 check(ok&&delivered==receipt.substr(0,receipt.size()-1)+"\r\n");
 check(!ot_console_begin_session());
 check(!ot_policy_send(receipt.data(),receipt.size()));
 }else{check(!ok&&!ot_console_healthy());auto count=accepted_bytes;check(!ot_policy_send(receipt.data(),receipt.size())&&accepted_bytes==count);}
 }
 std::cout<<"PASS "<<scenario<<"\n";
}
